database setup script

dockerdave00 opened the pull request:
This is the initial setup config for the database using a bash script and psql -c commands. Obviously this isn't needed in production, but I wanted to save it and so you could see how  I'm setting up.

dockerdave00 committed:
database setup script

dockerdave00 committed:
fixed potential bug, added SERIAL primary key feature

dockerdave00 committed:
database config file

dockerdave00 committed:
updates to current but not complete

dockerdave00 committed:
more dev on users:GET code

chrisghill commented on Dockerfile:
```diff
@@ -8,6 +8,8 @@ COPY requirements.txt .
 
 RUN pip install -r requirements.txt
 
+EXPOSE 5432
```
Not needed

dockerdave00 replied:
yea, in the MOST RECENT version of the Dockerfile, it was removed.

dockerdave00 commented:
I believe this is working as required.  Take a look and see if you agree.  If you want to run it, you can use the postgres_setup.sh file to setup the dateabase and redis and get them running prior to running the flask server.

dockerdave00 committed:
working homework7 as required for review

chrisghill commented on app.py:
```diff
@@ -69,4 +145,8 @@ def hello():
 if __name__ == '__main__':
+    success = initialize()
+    if not success:
+        print(f'Could not connect to services, HTTP: 503')
```
There wouldn't be a 503 here as there is no request.  This is just starting the webserver

dockerdave00 replied:
yea, I was curious how I would handle . I'll modify that to just a regular error statement.

chrisghill commented on app.py:
```diff
@@ -1,14 +1,92 @@
+            if id:
+                c.execute("SELECT * FROM users WHERE id = %s;", (id,))
+                record = c.fetchmany()
+            elif name and not id:
```
you don't even need the `and not id` here as its covered by the previous `if`, and this statement is in an `elif`

chrisghill commented on app.py:
```diff
@@ -1,14 +1,92 @@
+            elif name and not id:
+                c.execute("SELECT * FROM users WHERE name = %s;", (name,))
+                record = c.fetchmany()
+            elif not id and not name:
```
this can just be an `else` I think

dockerdave00 replied:
acknowledged. I didn't want to leave it to chance, but its easier to must make it an "else".

chrisghill commented on app.py:
```diff
@@ -1,14 +1,92 @@
+        try:
+            if id:
+                c.execute("SELECT * FROM users WHERE id = %s;", (id,))
+                record = c.fetchmany()
```
This can be a `fetchone()` as there should only be one record

dockerdave00 replied:
acknowledged.

chrisghill commented on app.py:
```diff
@@ -1,14 +1,92 @@
+                c.execute("SELECT * FROM users")
+                record = c.fetchall()
+        except (Exception) as e:
+            return 'Could not connect to database', 503
```
You're already connected to the DB, this would be a query issue

chrisghill commented on app.py:
```diff
@@ -1,14 +1,92 @@
+            return 'JSON POST failure', 400
+        else:
+            columns = d.keys()
+            values = [d[column] for column in columns]
```
Careful, you're not sanitizing inputs here.  Perhaps its fine in this case, but what if there was a column I *wasn't* supposed to be able to modify via a POST request, but I had it in the json.  It would get set based on this.

This is actually a really clever way to do it, and I like it.  But its currently not safe.  You need to make sure you *only* set `name`, `address` and `phone`

chrisghill replied:
As a side note, this is the sort of shit hackers look for when they hit endpoints.  They intentionally do shit to see if there are unsanitized DB queries.

dockerdave00 replied:
Yea, it may be clever, but its not mine. you're assuming I have intelligence wrapped up in how this works.  I "got it running" and figured it was probably a cluster, so let's discuss a better way to do it.

chrisghill commented on app.py:
```diff
@@ -1,14 +1,92 @@
+        if not record:
+            return_string = ""
+        else:
+            return_string = ""
```
Instead of building your own json string here, I would recommend you drop everything into a dict, and then convert it to json using the builtin library.  This ensures you get valid json.
https://www.geeksforgeeks.org/how-to-convert-python-dictionary-to-json/

dockerdave00 commented:
Updated json output on GET to list of tuples, list of dicts, and then json dump.  I also added some formatting on the return so its easier to read.

dockerdave00 committed:
updated get to conv tuples to dict to json

dockerdave00 commented:
Updated GET/POST logic:
- Found problems trying to iterate through "id" , so created logic to fix issue
- Updated POST logic to get only required fields and output error if one or all don't exist
- Added cursor close commands at appropriate positions

Update postgres DB creation logic:
- added volume information to RUN line and verified it creates remote volume
-  *** Still having problems... If I stop database and try to restart it errors because volume is part of old container. if I remove container, it remakes database in volume and wipes out data. Database is not persistent at this point.

dockerdave00 committed:
fix GET/POST logic, create persistent volume mounts

dockerdave00 committed:
update to add commit command for postgres

dockerdave00 commented:
Updated for final review.

dockerdave00 committed:
new docker-compose.yaml and postgres_setup.sh, remove Dockerfile, updates to init.sql app.py

chrisghill commented on app.py:
```diff
@@ -1,14 +1,88 @@
 from flask import Flask, request, render_template
 import redis
+import psycopg2
+from psycopg2.extensions import AsIs
```
This isn't used anymore, is it?

chrisghill commented on Dockerfile:
```diff
@@ -1,15 +0,0 @@
-  
```
We still need this file.  Keeping a Dockerfile at the root of a repo is pretty standard.

chrisghill commented on app.py:
```diff
@@ -1,14 +1,88 @@
+                list_of_dict = [dict(zip(keys, values)) for values in record]
+            json_out = json.dumps(list_of_dict)
+            c.close()
+            return json_out.replace("}, {", "},\n {") + '\n'
```
This makes the output more human readable, but it adds unnecessary bytes to the response. You wouldn't do this in a production system.  The more condense the response, the better.

chrisghill commented on app.py:
```diff
@@ -1,14 +1,88 @@
+                c.execute(insert_statement, data)
+                id_of_new_row = c.fetchone()[0]
+            except (Exception) as e:
+                c.close()
```
I would explicitly call `c.rollback()` here before closing the cursor to make it clear we are undoing all of our changes.

chrisghill commented on app.py:
```diff
@@ -69,4 +141,11 @@ def hello():
+
     app.run(debug=True, host='0.0.0.0')
+
+    p.close()
```
Only closing postgres here, not redis

dockerdave00 committed:
adding dockerfile
